internal/cms/signed.go

```diff
@@ -0,0 +1,313 @@
+	for _, cert := range verifiedSignerMap {
+		verifiedSigningCertificates = append(verifiedSigningCertificates, cert)
+	}
+	return verifiedSigningCertificates, nil
+}
+
+// VerifySigner verifies the signerInfo against the user specified signingCertificate.
+//
+// This function should be used when:
+//
+// 1. The certificates field of d is missing. This function allows the caller to provide
+// a signing certificate to verify the signerInfo.
+//
+// 2. The caller doesn't trust the signer identifier (unsigned field) of signerInfo
+// to identify signing certificate. This function allows such caller to use their trusted
+// signing certificate.
+//
+// Note: the intermediate certificates (if any) and root certificates in the verify
+// options MUST be set by the caller. The certificates field of d is not used in this function.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) VerifySigner(ctx context.Context, signerInfo *SignerInfo, signingCertificate *x509.Certificate, opts x509.VerifyOptions) (*x509.Certificate, error) {
+	if signerInfo == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signer info is required"}
+	}
+
+	if signingCertificate == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signing certificate is required"}
+	}
+
+	if signerInfo.Version != 1 {
+		// Only IssuerAndSerialNumber is supported currently
+		return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+	}
+
+	return d.verify(signerInfo, signingCertificate, &opts)
+}
+
+// verify verifies the trust in a top-down manner.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verify(signerInfo *SignerInfo, cert *x509.Certificate, opts *x509.VerifyOptions) (*x509.Certificate, error) {
+	// verify signer certificate
+	certChains, err := cert.Verify(*opts)
+	if err != nil {
+		return cert, VerificationError{Detail: err}
+	}
+
+	// verify signature
+	if err := d.verifySignature(signerInfo, cert); err != nil {
+		return nil, err
+	}
+
+	// verify attribute
+	return cert, d.verifySignedAttributes(signerInfo, certChains)
+}
+
+// verifySignature verifies the signature with a trusted certificate.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignature(signerInfo *SignerInfo, cert *x509.Certificate) error {
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signerInfo.DigestAlgorithm.Algorithm,
+		signerInfo.SignatureAlgorithm.Algorithm,
+	)
```

Comment:
I actually have a question on `algorithm` here, it is used in `cert.CheckSignature` as the `SignatureAlgorithm`. Can we get the `SignatureAlgorithm` from the cert's public key `type` and `size` like what we did in `notation-core-go`? See [ExtractKeySpec](https://github.com/notaryproject/notation-core-go/blob/2bc927beb626d46bb434215af243d49d629b13d8/signature/algorithm.go#L70) and [SignatureAlgorithm](https://github.com/notaryproject/notation-core-go/blob/2bc927beb626d46bb434215af243d49d629b13d8/signature/algorithm.go#L103).

Reply:
I believe we cannot do that for the following reasons:

1. Notation-core-go retrieves the signature algorithm from the certificate based on a fixed mapping. However, this approach is specific to the Notation's definition. We assume that a key with RSA 2048 should always sign with RSASSA-PSS with SHA-256, but in reality, an RSA 2048 key can also sign using RSASSA-PSS with SHA-512. Therefore, we cannot apply the same logic to the CMS package.
2. The `certificates` field is optional, so we should not rely on an optional field to obtain the signature algorithm.
3. The `digestAlgorithm` (`DigestAlgorithmIdentifier`) and `signatureAlgorithm` (`SignatureAlgorithmIdentifier`) in the `SignerInfo` are specifically designed fields in RFC 5652 to determine the signature algorithm, so we should use them directly.

Reply:
Question: from `oid.ToSignatureAlgorithm`, why are we not supporting `x509.SHA256WithRSAPSS`, `x509.SHA384WithRSAPSS`, and `x509.SHA512WithRSAPSS`? (https://pkg.go.dev/crypto/x509#SignatureAlgorithm)

Reply:
Updated.